### PR TITLE
Implement date_part for intervals

### PR DESCRIPTION
**Pull request by nrc**
# Which issue does this PR close?

Addresses https://github.com/apache/datafusion/issues/6327

Mentioned in #4969.

Closes https://github.com/apache/arrow-rs/issues/6113

# Rationale for this change
 
Will support using `extract` with intervals in DataFusion. It's also desirable functionality in Arrow for it's own sake, although I don't have a directly motivating use case.

# Are there any user-facing changes?

The `date_part` function will now work with intervals

### CODE REVIEW

**Comment by nrc** [comment]
https://github.com/apache/datafusion/pull/11501 demonstrates how this change supports `extract` on intervals in DataFusion.

**Comment by nrc** [comment]
@alamb @Jefffrey would either of you be able to review this PR please? (Sorry if I'm missing process for requesting a review or you're not the best folk, it's my first Arrow PR).

File: arrow-arith/src/temporal.rs [hunk 1]
```diff
@@ -1500,4 +1595,203 @@ mod tests {
...
+        assert_eq!(1, actual.value(1));
+        assert_eq!(5, actual.value(2));
+
+        let actual = date_part(&input, DatePart::Day).unwrap();
+        let actual = actual.as_primitive::<Int32Type>();
+        assert_eq!(0, actual.value(0));
+        assert_eq!(10, actual.value(1));
+        assert_eq!(35, actual.value(2));
+
+        // Times follow from nanos, but are not affected by months or dats.
```
Comment (alamb) [inline arrow-arith/src/temporal.rs:1731]:
```suggestion
        // Times follow from nanos, but are not affected by months or days.
```

**Review by alamb — APPROVED** [review]
Thank you @nrc  -- I think this PR looks really nice . Sorry for the delay in reviewing.

**Comment by alamb** [comment]
I took the liberty of merging up from main to pick up https://github.com/apache/arrow-rs/pull/6116

**Comment by alamb** [comment]
Thanks again @nrc

File: arrow-arith/src/temporal.rs (hunk 1 above)
Comment (alamb) [reply arrow-arith/src/temporal.rs:1731]: Proposed to fix in https://github.com/apache/arrow-rs/pull/6140

**Pull request merged**